### PR TITLE
feat(code-snippet): Add code snippet component

### PR DESCRIPTION
**Pull request by jpveooys**
Depends on #21

This adds the code snippet component.

This includes syntax highlighting (based on GOV.UK colours) and a copy code button.

The copy code button is not shown when JavaScript is unavailable, or the browser doesn't support `navigator.clipboard.writeText()` (i.e. old browsers).

Note that Playwright doesn't support testing the clipboard at the moment (it needs to be isolated since tests run in parallel) – had to test that using vitest and user-event.

@adamledwards Could you check how it looks on a Mac (in Safari and Chrome)?

# Demo (slightly out of date)

[Screencast from 19-01-23 13:56:20.webm](https://user-images.githubusercontent.com/66470099/213461055-f9834a53-dc57-45b9-9638-054804d5f7d6.webm)

### CODE REVIEW

File: src/site/_includes/components/code-snippet/macro.njk [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{%- macro guidanceCodeSnippet(params) -%}
+
+<div class="guidance-code-snippet">
+<button class="guidance-copy-button js-guidance-copy-button" data-code="{{ (params.contents or caller()) | dump }}">Copy code</button>
+{%- highlight params.language -%}
+{{ params.contents or caller() }}
```
Comment (jpveooys) [inline src/site/_includes/components/code-snippet/macro.njk:6]: Unsure at the moment which one will be used, so it allows both for now.

File: src/site/index.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
 // @ts-expect-error @moduk/frontend does not yet support types
 import * as mod from '@moduk/frontend/client'
+import { initCopyButtons } from './_includes/components/code-snippet'
```
Comment (jpveooys) [inline src/site/index.ts:3]: To discuss: I thought initCopyButtons was a meaningful name here. Also, I know sometimes linting rules say to use it, but I don't like default exports so much as they can lead to inconsistent names and you have to change it if you add another export.

File: src/site/components/back-link.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+---
+layout: layouts/markdown-content-page.njk
+title: Back link
+tags: components
```
Comment (jpveooys) [inline src/site/components/back-link.md:4]: Will probably need to do something to automatically sort these ones alphabetically.

File: src/site/_includes/global/prismjs-theme.scss [hunk 4]
```diff
@@ -0,0 +1,129 @@
+@use 'sass:color';
```
Comment (jpveooys) [inline src/site/_includes/global/prismjs-theme.scss:1]: Custom theme for prism.js with GOV.UK colours – had to darken some for contrast (against the grey background).

File: e2e/tests/pages/components/back-link.test.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      test('the outer box matches the screenshot on focus', async ({ page }) => {
+        const codeSnippet = page.locator('.guidance-code-snippet').first()
+        await codeSnippet.locator('pre').focus()
+
+        await expect(codeSnippet).toHaveScreenshot('copy-snippet-box-focus.png')
+      })
+
+      test('the copy button matches the screenshot on focus', async ({ browserName, page }) => {
+        // https://github.com/microsoft/playwright/issues/18901
+        test.skip(browserName === 'webkit', 'WebKit does not support navigator.clipboard on Linux')
```
Comment (jpveooys) [inline e2e/tests/pages/components/back-link.test.ts:23]: Had to skip this in webkit as per the comment – assume (or hope) it works in Safari running on an actual Mac though.

File: src/site/index.ts (hunk 2 above)
Comment (adamledwards) [reply src/site/index.ts:3]: I agree, `navigation()` and `copyButtons()` don't really mean anything. I will update in #23

File: src/site/components/back-link.md (hunk 3 above)
Comment (adamledwards) [reply src/site/components/back-link.md:4]: I had some issues with this in #23 I ordered them for testing.  I think if get-started and component have different sorting rules we will have to add a this 
https://www.11ty.dev/docs/collections/#advanced-custom-filtering-and-sorting

File: e2e/tests/pages/components/back-link.test.ts (hunk 5 above)
Comment (adamledwards) [reply e2e/tests/pages/components/back-link.test.ts:23]: Tested on Safari Version 16.1 works fine

File: src/site/index.ts (hunk 2 above)
Comment (jpveooys) [reply src/site/index.ts:3]: Cool, I will probably rename this one to `initCodeSnippets` also for consistency.

File: src/site/components/back-link.md (hunk 3 above)
Comment (jpveooys) [reply src/site/components/back-link.md:4]: Yeah I think we can sort alphabetically as a tie-break when the orders are equal (to make it one rule). But could centralise that one way or another (custom filter another option).

**Pull request merged**